### PR TITLE
feat(web): send cookies in grpc metadata

### PR DESCRIPTION
**Pull request by robot9706**
- Removed `accessedBy` field from the Crux gRPC API.
- Crux-UI sends cookies in gRPC Metadata which is then used to get the user ID.
- HTTP API uses user ID from the JWT.

### CODE REVIEW

File: web/crux/src/interceptors/grpc.user.interceptor.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * gRPC request user interceptor. Injects the userID into the metadata
+ * based on the cookie from the request.
+ *
+ * Disclaimer: Works only with gRPC controllers.
+ */
+@Injectable()
+export default class GrpcUserInterceptor implements NestInterceptor {
+  private readonly logger = new Logger(GrpcUserInterceptor.name)
```
Comment (polaroi8d) [inline web/crux/src/interceptors/grpc.user.interceptor.ts:34]: No using of the `logger`.

File: web/crux/src/services/kratos.service.ts [hunk 2]
```diff
@@ -1,15 +1,18 @@
...
-import { Configuration, Identity, IdentityApi, Session } from '@ory/kratos-client'
+import { Configuration, FrontendApi, Identity, IdentityApi, Session } from '@ory/kratos-client'
 import { randomUUID } from 'crypto'
 import { IdentityTraits, KratosInvitation, KRATOS_IDENTITY_SCHEMA } from 'src/shared/models'
 
 @Injectable()
 export default class KratosService {
   private kratos: IdentityApi
 
+  private what: FrontendApi
```
Comment (polaroi8d) [inline web/crux/src/services/kratos.service.ts:11]: `what`?

File: web/crux/src/interceptors/grpc.user.interceptor.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import KratosService from 'src/services/kratos.service'
+
+const DISABLE_ACCESSED_BY_METADATA = 'disable-accessed-by-metadata'
+
+export const DisableAccessedByMetadata = () => SetMetadata(DISABLE_ACCESSED_BY_METADATA, true)
+
+export const setAccessedByFromCookie = async (metadata: Metadata, kratos: KratosService) => {
+  const cookie = metadata.getMap().cookie as string
+  if (!cookie) {
+    throw new Error('Cookie missing')
```
Comment (polaroi8d) [inline web/crux/src/interceptors/grpc.user.interceptor.ts:14]: We can use here the `NotFoundException`, no?

File: web/crux/src/interceptors/grpc.user.interceptor.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  }
+
+  const session = await kratos.getSessionByCookie(cookie)
+
+  metadata.add('accessedby', session.identity.id)
```
Comment (polaroi8d) [inline web/crux/src/interceptors/grpc.user.interceptor.ts:19]: `accessedby` withouth camelCase is okey? Otherhand use const as DISABLED_ACCESSED... would be better, IMHO.

File: web/crux/src/interceptors/helper.interceptor.ts [hunk 5]
```diff
@@ -14,17 +14,19 @@ type GrpcCallLog = {
...
 export default class InterceptorGrpcHelperProvider {
   mapToGrpcObject(context: ExecutionContext): GrpcCallLog {
-    const request = context.getArgByIndex<AccessRequest>(0)
+    const request = context.getArgByIndex(0)
     const surfaceCall = context.getArgByIndex<ServerSurfaceCall>(2)
     return this.mapServerCallToGrpcLog(request, surfaceCall)
   }
 
-  mapServerCallToGrpcLog(request: AccessRequest, surfaceCall: ServerSurfaceCall): GrpcCallLog {
+  mapServerCallToGrpcLog(request: any, surfaceCall: ServerSurfaceCall): GrpcCallLog {
```
Comment (polaroi8d) [inline web/crux/src/interceptors/helper.interceptor.ts:22]: Can we avoid to using the `any`?

File: web/crux/src/shared/user-access.guard.ts [hunk 6]
```diff
@@ -20,19 +25,22 @@ export default abstract class UserAccessGuard implements CanActivate {
...
-
     if (!request.id) {
       return true
     }
 
-    return this.canActivateWithIdRequest(request)
+    const metadata = context.getArgByIndex<Metadata>(1)
+    const cookie = metadata.getMap().cookie as string
+    if (!cookie) {
+      throw new Error('Cookie missing')
```
Comment (polaroi8d) [inline web/crux/src/shared/user-access.guard.ts:35]: Throw the deleted Exception or NotFound would be better. 

```throw new InvalidArgumentException({
        message: 'Missing property',
        property: 'accessedBy',
      })```

File: web/crux/src/decorators/grpc.pipe.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import { Metadata } from '@grpc/grpc-js'
```
Comment (polaroi8d) [inline web/crux/src/decorators/grpc.pipe.ts:1]: Misleading the decorator folder, its a pipe, no?

File: web/crux/src/decorators/jwt-user.decorator.ts [hunk 8]
```diff
@@ -0,0 +1,8 @@
+import { createParamDecorator, ExecutionContext } from '@nestjs/common'
+
+const JWTUser = createParamDecorator((_: any, context: ExecutionContext) => {
```
Comment (polaroi8d) [inline web/crux/src/decorators/jwt-user.decorator.ts:3]: any? 🌷

**Review by polaroi8d — CHANGES_REQUESTED** [review]

File: web/crux/src/interceptors/grpc.user.interceptor.ts (hunk 3 above)
Comment (m8vago) [reply web/crux/src/interceptors/grpc.user.interceptor.ts:14]: I think an Unauthorized would be appropiate

File: web/crux/src/interceptors/helper.interceptor.ts (hunk 5 above)
Comment (m8vago) [reply web/crux/src/interceptors/helper.interceptor.ts:22]: As the AccessRequest was removed, we can not assume any other type than `any`

File: web/crux/src/shared/user-access.guard.ts (hunk 6 above)
Comment (m8vago) [reply web/crux/src/shared/user-access.guard.ts:35]: Unauthorized please

File: web/crux/src/interceptors/grpc.user.interceptor.ts (hunk 4 above)
Comment (robot9706) [reply web/crux/src/interceptors/grpc.user.interceptor.ts:19]: For some reason metadata keys are lowercased by default, so even if I used `accessedBy` when trying to get the key it only works with `accessedby`.

Putting it into a const is a good idea though 👍

File: web/crux-ui/src/components/products/versions/deployments/deployment-view-list.tsx [hunk 9]
```diff
@@ -35,7 +35,7 @@ const DeploymentViewList = (props: DeploymentViewListProps) => {
   ]
 
   const itemTemplate = (item: Instance) => [
-    item.image.name,
+    item.image.config.name,
```
Comment (m8vago) [inline web/crux-ui/src/components/products/versions/deployments/deployment-view-list.tsx:38]: we can't be sure it has a config

File: web/crux/src/decorators/jwt-user.decorator.ts (hunk 8 above)
Comment (robot9706) [reply web/crux/src/decorators/jwt-user.decorator.ts:3]: Yes because the request body type is unknown. I could maybe do some generic magic but the request body is not needed anyways.

File: web/crux/src/app/registry/registry.http.controller.ts [hunk 10]
```diff
@@ -20,7 +22,7 @@ export default class RegistryHttpController {
   @ApiBody({ type: AccessRequestDto })
   @ApiOkResponse({ type: RegistryListResponseDto })
   @AuditLogLevel('disabled')
-  async getRegistries(@Body() request: AccessRequest): Promise<RegistryListResponse> {
-    return await this.service.getRegistries(request)
+  async getRegistries(@Body() request: Empty, @JWTUser() accessedBy: string): Promise<RegistryListResponse> {
```
Comment (m8vago) [inline web/crux/src/app/registry/registry.http.controller.ts:25]: we don't need the empty here

**Review by m8vago — APPROVED** [review]

File: web/crux/src/shared/user-access.guard.ts [hunk 11]
```diff
@@ -1,38 +1,41 @@
...
-        message: 'Missing property',
-        property: 'accessedBy',
-      })
+    const metadata = context.getArgByIndex<Metadata>(1)
+    const cookie = metadata.getMap().cookie as string
+    if (!cookie) {
+      throw new UnauthorizedException()
     }
 
-    if (!request.id) {
```
Comment (m8vago) [inline web/crux/src/shared/user-access.guard.ts:30]: Why did you move this to the concrete access guards?

**Review by m8vago — APPROVED** [review]
